src/FwdState.cc

```diff
@@ -964,6 +938,47 @@ FwdState::connectStart()
     AsyncJob::Start(cs);
 }
 
+/// send request on an existing connection dedicated to the requesting client
+void
+FwdState::usePinned()
+{
+    // we only handle pinned destinations; others are handled by connectStart()
+    assert(!serverDestinations.empty());
+    assert(!serverDestinations[0]);
+
+    const auto connManager = request->pinnedConnection();
+    debugs(17, 7, "connection manager: " << connManager);
+
+    // the client connection may close while we get here, nullifying connManager
+    const auto temp = connManager ? connManager->borrowPinnedConnection(request) : nullptr;
+    debugs(17, 5, "connection: " << temp);
+
+    // the previously pinned idle peer connection may get closed (by the peer)
+    if (!Comm::IsConnOpen(temp)) {
+        serverConn = nullptr;
```

Comment:
We left this pinned-connection-error code path without a syncHierNote() call, right? The comments above the syncHierNote() call in startConnection() sound compelling. Should not we add that call here as well, for similar reasons?

Reply:
Yes we need it.

A  "syncHierNote(temp);" call is enough in the case of temp is not open (or it is nil).

Alternatively we can call syncHierNote(temp) immediately after this variable is set and make new function to set qos values (the part of job of syncWithServerConn) if the temp is open.
But I think the first is better.

Reply:
> Yes we need it. A "syncHierNote(temp);" call is enough

Please add.

----

P.S. I am reposing this comment here, in the right change request, after GitHub screwed up and added my email response to the general PR discussion area.

Reply:
Please note that you have [removed](https://github.com/squid-cache/squid/pull/351#event-2086674142) `S-waiting-for-author` without addressing this change request (or without pushing the corresponding commit to GitHub).

Reply:
addressed at 144ab339